# Treue Technologies (Internship - Data Science) [Ansh Pandey]

In [1]:
pip install -U imbalanced-learn

## Task (1) : Bank Customer Churn Prediction (ML Model)

### Objective = Train a machine learning model to predict whether a bank customer will churn (leave the bank) or not based on various customer attributes and banking behavior. The goal is to create a model that can accurately identify customers who are likely to churn, enabling proactive retention strategies. 

#### i)  Importing Necessary Libraries

In [23]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix, roc_auc_score
from imblearn.over_sampling import SMOTE
from sklearn.model_selection import GridSearchCV



#### ii) Loading and Exploring the Data

In [24]:
data = pd.read_csv("C:\\Users\\asus\\Desktop\\Data Science or Analytics_Ansh\\_Bank Customer Churn Prediction\Churn_Modelling.csv")

In [25]:
print(data)

      RowNumber  CustomerId    Surname  CreditScore Geography  Gender  Age  \
0             1    15634602   Hargrave          619    France  Female   42   
1             2    15647311       Hill          608     Spain  Female   41   
2             3    15619304       Onio          502    France  Female   42   
3             4    15701354       Boni          699    France  Female   39   
4             5    15737888   Mitchell          850     Spain  Female   43   
...         ...         ...        ...          ...       ...     ...  ...   
9995       9996    15606229   Obijiaku          771    France    Male   39   
9996       9997    15569892  Johnstone          516    France    Male   35   
9997       9998    15584532        Liu          709    France  Female   36   
9998       9999    15682355  Sabbatini          772   Germany    Male   42   
9999      10000    15628319     Walker          792    France  Female   28   

      Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMemb

In [26]:
data

RowNumber  CustomerId    Surname  CreditScore Geography  Gender  Age  \
0             1    15634602   Hargrave          619    France  Female   42   
1             2    15647311       Hill          608     Spain  Female   41   
2             3    15619304       Onio          502    France  Female   42   
3             4    15701354       Boni          699    France  Female   39   
4             5    15737888   Mitchell          850     Spain  Female   43   
...         ...         ...        ...          ...       ...     ...  ...   
9995       9996    15606229   Obijiaku          771    France    Male   39   
9996       9997    15569892  Johnstone          516    France    Male   35   
9997       9998    15584532        Liu          709    France  Female   36   
9998       9999    15682355  Sabbatini          772   Germany    Male   42   
9999      10000    15628319     Walker          792    France  Female   28   

      Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0          2       0.00              1          1               1   
1          1   83807.86              1          0               1   
2          8  159660.80              3          1               0   
3          1       0.00              2          0               0   
4          2  125510.82              1          1               1   
...      ...        ...            ...        ...             ...   
9995       5       0.00              2          1               0   
9996      10   57369.61              1          1               1   
9997       7       0.00              1          0               1   
9998       3   75075.31              2          1               0   
9999       4  130142.79              1          1               0   

      EstimatedSalary  Exited  
0           101348.88       1  
1           112542.58       0  
2           113931.57       1  
3            93826.63       0  
4            79084.10       0  
...               ...     ...  
9995         96270.64       0  
9996        101699.77       0  
9997         42085.58       1  
9998         92888.52       1  
9999         38190.78       0  

[10000 rows x 14 columns]

In [27]:
data.head()

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0       2       0.00              1          1               1   
1       1   83807.86              1          0               1   
2       8  159660.80              3          1               0   
3       1       0.00              2          0               0   
4       2  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0

In [28]:
print(data.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           10000 non-null  object 
 6   Age              10000 non-null  int64  
 7   Tenure           10000 non-null  int64  
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(2), int64(9), object(3)
memory usage: 1.1+ MB
None


In [29]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           10000 non-null  object 
 6   Age              10000 non-null  int64  
 7   Tenure           10000 non-null  int64  
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(2), int64(9), object(3)
memory usage: 1.1+ MB


In [30]:
data.shape

(10000, 14)

In [31]:
data.size

140000

In [32]:
data.info

<bound method DataFrame.info of       RowNumber  CustomerId    Surname  CreditScore Geography  Gender  Age  \
0             1    15634602   Hargrave          619    France  Female   42   
1             2    15647311       Hill          608     Spain  Female   41   
2             3    15619304       Onio          502    France  Female   42   
3             4    15701354       Boni          699    France  Female   39   
4             5    15737888   Mitchell          850     Spain  Female   43   
...         ...         ...        ...          ...       ...     ...  ...   
9995       9996    15606229   Obijiaku          771    France    Male   39   
9996       9997    15569892  Johnstone          516    France    Male   35   
9997       9998    15584532        Liu          709    France  Female   36   
9998       9999    15682355  Sabbatini          772   Germany    Male   42   
9999      10000    15628319     Walker          792    France  Female   28   

      Tenure    Balance  NumOfP

In [33]:
data.columns

Index(['RowNumber', 'CustomerId', 'Surname', 'CreditScore', 'Geography',
       'Gender', 'Age', 'Tenure', 'Balance', 'NumOfProducts', 'HasCrCard',
       'IsActiveMember', 'EstimatedSalary', 'Exited'],
      dtype='object')

#### iii) Data Preprocessing

In [34]:
# Seperating data into numeric and categorical features
numeric_features = ['CreditScore', 'Age', 'Balance', 'NumOfProducts', 'Tenure', 'HasCrCard', 'IsActiveMember', 'EstimatedSalary']
categorical_features = ['Geography', 'Gender']

X_numeric = data[numeric_features]
X_categorical = data[categorical_features]

# Creating transformers for preprocessing
numeric_transformer = StandardScaler()
categorical_transformer = OneHotEncoder(drop='first', sparse_output= False)

# Fit and transform the transformers on the respective data
X_numeric = numeric_transformer.fit_transform(X_numeric)
X_categorical = categorical_transformer.fit_transform(X_categorical)

# combine the numeric and categorical features back togather
X = np.hstack((X_numeric, X_categorical))


In [35]:
# All Columns Name 
data.columns

Index(['RowNumber', 'CustomerId', 'Surname', 'CreditScore', 'Geography',
       'Gender', 'Age', 'Tenure', 'Balance', 'NumOfProducts', 'HasCrCard',
       'IsActiveMember', 'EstimatedSalary', 'Exited'],
      dtype='object')

In [36]:
# Defined the columns to be scaled
numeric_features = ['Age', 'Balance', 'CreditScore', 'NumOfProducts']

# Created a preprocessor that scales numeric features
numeric_transformer = Pipeline(steps=[
    ('scaler', StandardScaler())
])

# Combined the transformers using ColumnTransformer
preprocessor = ColumnTransformer(
    transformers=[
        ('num', numeric_transformer, numeric_features)
        # You can add more transformers for other feature types if needed
    ])

# Now I can using the preprocessor to transform my data
X = data.drop(['CustomerId', 'Exited'], axis=1)
X = preprocessor.fit_transform(X)

In [37]:
#data = data.dropna()
#data = pd.get_dummies(data, columns = ['Geography', 'Gender'])

### Feature Scaling
#scaler = StandardScaler()
#X= data.drop(['CustomerId', 'Exited'], axis=1)
#X= preprocessor.fit_transform(X)

y = data['Exited']

In [38]:
import pandas as pd
from sklearn.preprocessing import OneHotEncoder

categorical_columns = ['Geography', 'Gender']

# creating a dataframe containing only the categorical columns
data_categorical = data[categorical_columns]

# Initialize the OneHotEncoder
encoder = OneHotEncoder(drop= 'first', sparse_output= False)

# Fitting and transforming the encoder on the categorical data
encoded_data = encoder.fit_transform(data_categorical)

# Get the feature names for the one-hot encoded columns
encoded_columns = encoder.get_feature_names_out(input_features = categorical_columns)

# Creating a Dataframe from the encoded data with appropriate column names
encoded_df = pd.DataFrame(encoded_data, columns = encoder.get_feature_names_out(categorical_columns))

# Dropping the original categorical columns from your original DataFrame
data = pd.concat([data, encoded_df], axis = 1)

# Now, I am dropping the original categorical columns if needed 
data.drop(categorical_columns, axis=1, inplace= True)


In [39]:
# Creating a  column transformer to apply transformers to the appropriate features
preprocessor = ColumnTransformer(
    transformers=[
        ('num', numeric_transformer, numeric_features),
        ('cat', categorical_transformer, categorical_features)
    ])

#### iv) Split data and handle class imbalance

In [40]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)

# Applying SMOTE to address class imbalance
smote = SMOTE(sampling_strategy = 'auto', random_state = 0)
X_train_resampled, y_train_resampled = smote.fit_resample(X_train, y_train)


#### v) Model Selection and Hyperparameter Tuning

In [42]:
# Defining the Random Forest classifier and parameter grid for hyperparameter tuning
param_grid = {
    'n_estimators': [100, 200, 300],
    'max_depth': [5,10,15],
    'min_samples_split': [2,5,10],
    'min_samples_leaf': [1,2,4]
}


rf_classifier = RandomForestClassifier(random_state = 0)

# Performing hyperparameter tuning using GridsearchCV
grid_search = GridSearchCV(estimator = rf_classifier, param_grid = param_grid, cv = 5, scoring = 'accuracy', n_jobs = -1)
grid_search.fit(X_train_resampled, y_train_resampled)

# Get the best hypeerparameters
best_params = grid_search.best_params_
best_estimator = grid_search.best_estimator_
print("Best Hyperparameters:", best_params)

# Traninng the Random Forest Model with the best hyperparameters
best_rf_classifier = RandomForestClassifier(**best_params, random_state = 42)
best_rf_classifier.fit(X_train_resampled, y_train_resampled)

Best Hyperparameters: {'max_depth': 15, 'min_samples_leaf': 1, 'min_samples_split': 2, 'n_estimators': 100}


RandomForestClassifier(max_depth=15, random_state=42)

#### vi) Model Evaluation

In [43]:
from sklearn.metrics import accuracy_score, roc_auc_score, confusion_matrix, classification_report

# Predict on the test set
y_pred = best_rf_classifier.predict(X_test)

# Assuming I have already defined y_test and y_pred

accuracy = accuracy_score(y_test, y_pred)
roc_auc = roc_auc_score(y_test, y_pred)
conf_matrix = confusion_matrix(y_test, y_pred)
class_report = classification_report(y_test, y_pred)

print(f'Accuracy: {accuracy}')
print(f'ROC AUC: {roc_auc}')
print('Confusion Matrix:')
print(conf_matrix)
print('Classification')

Accuracy: 0.7995
ROC AUC: 0.7296838112930067
Confusion Matrix:
[[1351  244]
 [ 157  248]]
Classification


#### vii) Interpretability and Feature Importance

In [47]:
# Getting feature importance from the trained model
feature_importances = best_rf_classifier.feature_importances_

# Check the length of feature_importances and columns
if len(feature_importances) == len(data.drop(['CustomerId','Exited'], axis = 1).columns):
    
    # Create the barplot
    plt.figure(figsize =(10, 6))
    sns.barplot(x = feature_importances, y = data.drop(['CustomerId', 'Exited'], axis = 1).columns)
    plt.xlabel('Feature Importance')
    plt.ylabel('Features')
else:
    print("Length mismatch between feature_importances and columns for plotting.")
    

Length mismatch between feature_importances and columns for plotting.


#### viii) Deployment and Continuous Monitoring

In [57]:
import joblib

# Train a machine learning model
model = RandomForestClassifier()
model.fit(X_train, y_train)

# Evaluate the model
y_pred = model.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print(f"Model Accuracy: {accuracy:.2f}")

# Save the trained model
joblib.dump(model, "churn_prediction_model.pkl")

# Deployment (simplified)
# In a real deployment, we'd use web frameworks, APIs, and cloud services for scalability.
# Here, we just load the model for simplicity.

loaded_model = joblib.load("churn_prediction_model.pkl")

# Continuous Monitoring (simplified)
# In practice, we would set up a monitoring system to retrain and update the model regularly.

# New data for prediction
new_data = pd.DataFrame({
    "Age": [35],
    "CreditScore": [750],
    "Tenure": [5],
    "Balance": [10000],
    #"NumOfProducts": [2],
    #"IsActiveMember": [1],
    #"NumOfChildren": [0]
})

# Make predictions on new data
new_predictions = loaded_model.predict(new_data)
print(f"Churn Prediction for New Customer: {'Churn' if new_predictions[0] == 1 else 'No Churn'}")


Model Accuracy: 0.82
Churn Prediction for New Customer: Churn


C:\Users\asus\anaconda3\Lib\site-packages\sklearn\base.py:458: UserWarning: X has feature names, but RandomForestClassifier was fitted without feature names
  warnings.warn(


## Conclusion

###  In conclusion, my Bank Customer Churn Prediction ML Model, with an accuracy of 82%, provides a valuable tool for banks to foresee and mitigate customer churn. This proactive approach can help banks retain valuable customers, reduce revenue loss, and enhance overall customer satisfaction. To further improve the model, ongoing monitoring and updates are essential, ensuring it continues to meet the evolving needs of the banking industry.
